tool.c

```diff
@@ -221,9 +222,105 @@ compdb(int argc, char *argv[])
 	return 0;
 }
 
+static void
+targets_rule(int argc, char *argv[])
+{
+	struct edge *e = NULL;
+	for (e = alledges; e; e = e->allnext) {
+		if (argc == 2) {
+			if (e->nin == 0)
+				continue;
+
+			if (!e->in[0]->gen)
+				printf("%s\n", e->in[0]->path->s);
+		}
+		else if (argc == 3) {
+			if (e->nout == 0)
+				continue;
+
+			if (strcmp(e->rule->name, argv[2]) == 0)
+				printf("%s\n", e->out[0]->path->s);
+		}
+	}
+}
+
+static void
+targets_depth(struct node **nodes, size_t nnodes, size_t depth, size_t indent)
+{
+	for (size_t i = 0; i < nnodes; ++i) {
+		if (nodes[i]->gen == NULL)
+			continue;
+
+		for (size_t j = 0; j < indent; ++j)
+			printf("  ");
+
+		printf("%s: %s\n", nodes[i]->path->s, nodes[i]->gen->rule->name);
+
+		if (depth > 1 || depth <= 0)
+			targets_depth(nodes[i]->gen->in, nodes[i]->gen->nin,
+					depth - 1, indent + 1);
+	}
+}
+
+static void
+targets_all(void)
+{
+	struct edge *e = NULL;
+	for (e = alledges; e; e = e->allnext) {
+		if (e->nout == 0)
+			continue;
+
+		printf("%s: %s\n", e->out[0]->path->s, e->rule->name);
+	}
+}
+
+static int
+targets(int argc, char *argv[])
+{
+	size_t depth = 1;
+	if (argc >= 2) {
+		const char *mode = argv[1];
+		if (strcmp(mode, "rule") == 0) {
+			targets_rule(argc, argv);
+			return 0;
+		} else if (strcmp(mode, "depth") == 0) {
+			if (argc > 1)
+				sscanf(argv[2], "%zu", &depth);
+		} else if (strcmp(mode, "all") == 0) {
+			targets_all();
+			return 0;
+		} else {
+			fprintf(stderr, "unknown target tool mode '%s'\n", mode);
+			return 2;
+		}
+	}
+
+	/* Find root nodes (node with 0 use) */
+	size_t size = 0;
+	struct node **nodes = NULL;
+	struct edge *e = NULL;
+	for (e = alledges; e; e = e->allnext) {
+		if (e->nout == 0)
+			continue;
+
+		for (size_t n = 0; n < e->nout; ++n) {
+			if (e->out[n]->nuse != 0)
+				continue;
+
+			nodes = xreallocarray(nodes, size + 1, sizeof(nodes[0]));
+			nodes[size] = e->out[n];
+			size += 1;
+		}
```

Comment:
Instead of allocating an array one by one here, then passing the result to `targets_depth`, can't you just replace this for loop with

```c
		targets_depth(e->out, e->nout, depth, 0);
```

?

Reply:
I completely missed this but yes, it can be done that way and can avoid reallocating the array. This is done in e8003ad